File: game/scripts/vscripts/abilities/heroes/nevermore/nevermore_mobility/nevermore_mobility.lua
nevermore_mobility = class({})
LinkLuaModifier( "modifier_nevermore_mobility_movement", "abilities/heroes/nevermore/nevermore_mobility/modifier_nevermore_mobility_movement", LUA_MODIFIER_MOTION_HORIZONTAL )
LinkLuaModifier( "modifier_nevermore_souls", "abilities/heroes/nevermore/nevermore_shared_modifiers/modifier_nevermore_souls", LUA_MODIFIER_MOTION_NONE )

function nevermore_mobility:OnCastPointEnd()
    local caster = self:GetCaster()
    self.origin = caster:GetOrigin()

	local damage = self:GetSpecialValueFor("ability_damage")
    local min_range = self:GetSpecialValueFor("min_range")
    local fading_slow_duration = self:GetSpecialValueFor("fading_slow_duration")
    local fading_slow_pct = self:GetSpecialValueFor("fading_slow_pct")
    local mana_gain_pct = self:GetSpecialValueFor("mana_gain_pct")
	local point = CalcPoint(self.origin, self:GetCursorPosition(), self:GetCastRange(Vector(0,0,0), nil), min_range)

    local direction = (point - self.origin):Normalized()
    local distance = (point - self.origin):Length2D()
    local speed = (distance/0.5)
	local basic_attack = caster:FindAbilityByName("nevermore_basic_attack")

    caster:RemoveModifierByName("modifier_generic_displacement")
    caster:AddNewModifier(
        caster, -- player source
        self, -- ability source
        "modifier_generic_displacement", -- modifier name
        {
            x = direction.x,
            y = direction.y,
            r = distance,
            speed = speed,
            peak = 0,
            colliding = 0,
            activity = 2,
        } -- kv
    )

	local offset = 100
	local hitbox = 150
	
    local projectile = {
        vSpawnOrigin = self.origin,
        fDistance = distance,
        fUniqueRadius = hitbox,
        Source = caster,
        vVelocity = direction * speed,
        UnitBehavior = PROJECTILES_NOTHING,
        TreeBehavior = PROJECTILES_NOTHING,
        WallBehavior = PROJECTILES_DESTROY,
        GroundBehavior = PROJECTILES_NOTHING,
        fGroundOffset = 0,
        bIsReflectable = false,
        bIsSlowable = false,
        UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not _self.Source:IsAlly(unit) end,
        OnUnitHit = function(_self, unit) 
			local counter = 0
			for k, v in pairs(_self.rehit) do
				counter = counter + 1
            end
            
            local damage_table = {
                victim = unit,
                attacker = _self.Source,
                damage = damage,
                damage_type = DAMAGE_TYPE_PURE,
            }
            ApplyDamage( damage_table )

            if _self.Source == caster and not unit:IsObstacle() then
                for i = 0, 1 do
                    caster:AddNewModifier(
                        caster,
                        basic_attack,
                        "modifier_nevermore_souls",
                        {}
                    )
                end
                
                if counter == 1 then
                    caster:GiveManaPercent(mana_gain_pct, unit)
                end
            end

            unit:AddNewModifier(_self.Source, self, "modifier_generic_fading_slow_new", { 
				duration = fading_slow_duration,
				max_slow_pct = fading_slow_pct
			})

            self:PlayEffectsOnImpact(unit)
        end,
        OnThinkBegin = function(_self, pos)
            local particle_cast = "particles/mod_units/heroes/hero_nevermore/nevermore_shadowraze.vpcf"
            local effect_cast = ParticleManager:CreateParticle( particle_cast, PATTACH_WORLDORIGIN, nil )
            ParticleManager:SetParticleControl( effect_cast, 0, pos - direction * 80)
            ParticleManager:SetParticleControl( effect_cast, 1, Vector( 250, 1, 1 ) )
            ParticleManager:ReleaseParticleIndex( effect_cast )
        end,
    }

    Projectiles:CreateProjectile(projectile)
    self:PlayEffectsOnCast()
end


function nevermore_mobility:OnDisplacementEnd()
    self:PlayEffectsOnDisplacementEnd()
end

function nevermore_mobility:OnDisplacementHalf()
    self:PlayEffectsOnDisplacementHalf()
end

--------------------------------------------------------------------------------
-- Graphics & sounds
function nevermore_mobility:PlayEffectsOnCast()
    EmitSoundOn("Hero_Nevermore.Death", self:GetCaster())
end

function nevermore_mobility:PlayEffectsOnDisplacementHalf()
    local particle_cast = "particles/econ/items/shadow_fiend/sf_desolation/sf_desolation_scratch.vpcf"
    local effect_cast = ParticleManager:CreateParticle( particle_cast, PATTACH_ABSORIGIN_FOLLOW, self:GetCaster())
    ParticleManager:SetParticleControl(effect_cast, 0, self:GetCaster():GetOrigin()) 
	ParticleManager:ReleaseParticleIndex( effect_cast )
end

function nevermore_mobility:PlayEffectsOnDisplacementEnd()
    local particle_cast = "particles/mod_units/heroes/hero_nevermore/nevermore_shadowraze.vpcf"
    local effect_cast = ParticleManager:CreateParticle( particle_cast, PATTACH_WORLDORIGIN, nil )
    ParticleManager:SetParticleControl( effect_cast, 0, self:GetCaster():GetOrigin())
    ParticleManager:SetParticleControl( effect_cast, 1, Vector( 250, 1, 1 ) )
    ParticleManager:ReleaseParticleIndex( effect_cast )
end

function nevermore_mobility:PlayEffectsOnImpact( hTarget )
	local projectile_name = "particles/units/heroes/hero_nevermore/nevermore_necro_souls.vpcf"
    EmitSoundOn("Hero_Nevermore.ProjectileImpact", hTarget)

	local info = {
		Target = self:GetCaster(),
		Source = hTarget,
		EffectName = projectile_name,
		iMoveSpeed = 400,
		vSourceLoc= hTarget:GetAbsOrigin(),                -- Optional
		bDodgeable = false,                                -- Optional
		bReplaceExisting = false,                         -- Optional
		flExpireTime = GameRules:GetGameTime() + 5,      -- Optional but recommended
		bProvidesVision = false,                           -- Optional
	}
	ProjectileManager:CreateTrackingProjectile(info)
end

if IsClient() then require("wrappers/abilities") end
Abilities.Initialize( 
	nevermore_mobility,
	{ activity = ACT_DOTA_RAZE_1, rate = 1.1 },
	{ movement_speed = 0 }
)

File: src/vscripts/abilities/heroes/nevermore/nevermore_mobility.ts
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { DisplacementParams, ModifierDisplacement, OnCollisionEvent } from "../../../modifiers/modifier_displacement";
import { ModifierFadingSlow } from "../../../modifiers/modifier_fading_slow";
import { clampPosition, direction2D, giveManaAndEnergyPercent, isGem, isObstacle } from "../../../util";
import { CustomAbility } from "../../framework/custom_ability";
import { CustomModifier } from "../../framework/custom_modifier";
import { ModifierNevermoreStacks } from "./nevermore_basic_attack";

@registerAbility("nevermore_mobility")
class NevermoreMobility extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_RAZE_1;
    }
    GetPlaybackRateOverride() {
        return 1.1;
    }
    GetCastingCrawl() {
        return 0;
    }

    OnSpellStart() {
        const origin = this.caster.GetAbsOrigin();
        const cursor = CustomAbilitiesLegacy.GetCursorPosition(this);
        const point = clampPosition(origin, cursor, {
            maxRange: this.GetCastRange(Vector(0, 0, 0), undefined),
            minRange: this.GetSpecialValueFor("min_range")
        });
        const direction = direction2D(origin, point);
        const distance = point.__sub(origin).Length2D();

        ModifierNevermoreMobility.apply(this.caster, this.caster, this, {
            x: direction.x,
            y: direction.y,
            distance: distance,
            speed: distance / 0.4,
            peak: 0,
            teamFilter: UnitTargetTeam.ENEMY
        });

        this.PlayEffectsOnCast();
    }

    PlayEffectsOnCast() {
        EmitSoundOn("Hero_Nevermore.Death", this.caster);
    }
}

@registerModifier({ customNameForI18n: "modifier_nevermore_mobility_displacement" })
class ModifierNevermoreMobility extends ModifierDisplacement {
    OnCreated(params: DisplacementParams) {
        super.OnCreated(params);

        if (IsServer()) {
            const particleId = ParticleManager.CreateParticle(
                "particles/econ/items/shadow_fiend/sf_desolation/sf_desolation_scratch.vpcf",
                ParticleAttachment.ABSORIGIN_FOLLOW,
                this.parent
            );
            ParticleManager.SetParticleControl(particleId, 0, this.parent.GetAbsOrigin());
            ParticleManager.ReleaseParticleIndex(particleId);
        }
    }

    OnDestroy() {
        super.OnDestroy();

        if (IsServer()) {
            this.parent.StartGestureWithPlaybackRate(GameActivity.DOTA_TELEPORT_END, 2.0);
            this.PlayEffectsOnDestroy();
        }
    }

    OnCollide(event: OnCollisionEvent) {
        if (IsServer() && event.collision == "unit") {
            for (const unit of event.units.filter(unit => !ModifierNevermoreMobilityMark.findOne(unit))) {
                ApplyDamage({
                    victim: unit,
                    attacker: this.parent,
                    damage: this.Value("ability_damage"),
                    damage_type: DamageTypes.PURE
                });

                if (!isObstacle(unit) && !isGem(unit)) {
                    giveManaAndEnergyPercent(this.parent, this.Value("mana_gain_pct"), true);

                    ModifierNevermoreStacks.findOne(this.caster)?.IncrementStackCount();
                    ModifierNevermoreStacks.findOne(this.caster)?.IncrementStackCount();

                    this.PlayEffectsOnImpact(unit);
                    this.PlayEffectsOnImpact(unit);
                }

                ModifierNevermoreMobilityMark.apply(unit, this.parent, this.ability, { duration: 1.0 });

                ModifierFadingSlow.apply(unit, this.parent, undefined, {
                    duration: this.Value("fading_slow_duration"),
                    maxSlowPct: this.Value("fading_slow_pct")
                });
            }
        }
    }

    // GetOnThinkCallback(){
    //     this.PlayEffectsOnDestroy()
    // }

    PlayEffectsOnDestroy() {
        const particleId = ParticleManager.CreateParticle(
            "particles/nevermore/nevermore_mobility.vpcf",
            ParticleAttachment.WORLDORIGIN,
            undefined
        );
        ParticleManager.SetParticleControl(particleId, 0, this.parent.GetAbsOrigin());
        ParticleManager.SetParticleControl(particleId, 1, Vector(250, 1, 1));
        ParticleManager.ReleaseParticleIndex(particleId);
    }

    PlayEffectsOnImpact(target: CDOTA_BaseNPC) {
        EmitSoundOn("Hero_Nevermore.ProjectileImpact", target);

        const info = {
            Target: this.parent,
            Source: target,
            EffectName: "particles/units/heroes/hero_nevermore/nevermore_necro_souls.vpcf",
            iMoveSpeed: 400,
            vSourceLoc: target.GetAbsOrigin(),
            bDodgeable: false,
            bReplaceExisting: false,
            flExpireTime: GameRules.GetGameTime() + 5,
            bProvidesVision: false
        };
        ProjectileManager.CreateTrackingProjectile(info);
    }

    DeclareFunctions() {
        return [ModifierFunction.OVERRIDE_ANIMATION, ModifierFunction.TRANSLATE_ACTIVITY_MODIFIERS];
    }

    GetOverrideAnimation() {
        return GameActivity.DOTA_FLAIL;
    }

    GetActivityTranslationModifiers() {
        return "forcestaff_friendly";
    }

    CheckState() {
        return {
            ...super.CheckState(),
            [ModifierState.NO_HEALTH_BAR]: true,
            [ModifierState.INVULNERABLE]: true,
            [ModifierState.OUT_OF_GAME]: true,
            [ModifierState.NO_UNIT_COLLISION]: true
        };
    }
}

@registerModifier({ customNameForI18n: "modifier_nevermore_mobility_hit" })
class ModifierNevermoreMobilityMark extends CustomModifier {
    IsPermanent() {
        return true;
    }

    IsHidden() {
        return true;
    }
}
